### PR TITLE
container: allow to iterate over container list

### PR DESCRIPTION
**Pull request by fyrchik**
Signed-off-by: Evgenii Stratonikov <evgeniy@morphbits.ru>

### CODE REVIEW

File: container/container_contract.go [hunk 1]
```diff
@@ -93,6 +95,32 @@ func _deploy(data interface{}, isUpdate bool) {
 	if isUpdate {
 		args := data.([]interface{})
 		common.CheckVersion(args[len(args)-1].(int))
+
```
Comment (roman-khimov) [inline container/container_contract.go:98]: There has to be some check for the current contract version here, if it's `>= some` this migration is not needed.

File: container/container_contract.go [hunk 2]
```diff
@@ -391,17 +419,24 @@ func Owner(containerID []byte) []byte {
...
-		if len(key) == containerIDSize {
-			count++
-		}
+		count++
 	}
 	return count
 }
 
+// Iterate iterates over all container IDs owned by the specified owner.
+func Iterate(owner []byte) iterator.Iterator {
```
Comment (roman-khimov) [inline container/container_contract.go:430]: `ContainersOf`? Frankly, I'm thinking about some NEP-11 compatibility for this contract.

File: container/container_contract.go [hunk 3]
```diff
@@ -391,17 +419,24 @@ func Owner(containerID []byte) []byte {
...
+	ctx := storage.GetReadOnlyContext()
+
+	if len(owner) == 0 {
+		return storage.Find(ctx, []byte(containerKeyPrefix), storage.KeysOnly|storage.RemovePrefix)
```
Comment (roman-khimov) [inline container/container_contract.go:434]: Wow, this differs a lot from `ValuesOnly` below, this shouldn't be done, we want iterators to have consistent values. If you need some different behavior just add another method, it's not hard to do that. And makes this one only accept correct owners (like `tokensOf` does).

File: container/container_contract.go [hunk 4]
```diff
@@ -62,6 +62,8 @@ const (
 
 	singleEstimatePrefix = "est"
 	estimateKeyPrefix    = "cnr"
+	containerKeyPrefix   = "cid"
+	ownerKeyPrefix       = "own"
```
Comment (roman-khimov) [inline container/container_contract.go:66]: We have some 3-letter prefixes already unfortunately, but I'd use 1-byte for new ones anyway.

File: container/container_contract.go [hunk 5]
```diff
@@ -93,6 +95,32 @@ func _deploy(data interface{}, isUpdate bool) {
...
+				value []byte
+			})
+			if len(item.key) == containerIDSize {
+				storage.Delete(ctx, item.key)
+				storage.Put(ctx, append([]byte(containerKeyPrefix), item.key...), item.value)
+			}
+		}
+
+		// Migrate owner-cid map.
+		it = storage.Find(ctx, []byte{}, storage.None)
```
Comment (roman-khimov) [inline container/container_contract.go:113]: Why two iterations over the DB?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: container/container_contract.go (hunk 1 above)
Comment (fyrchik) [reply container/container_contract.go:98]: Our current policy is that we remove this code after release and update "min updating from" versions in `common` package. This way we can update contract multiple times, but do not have complex in `_deploy`.

File: container/container_contract.go (hunk 4 above)
Comment (fyrchik) [reply container/container_contract.go:66]: Fixed. Initially I wanted to use 'c' as a prefix which would conflict with `cnr`, but nobody cares really.

File: container/container_contract.go (hunk 5 above)
Comment (fyrchik) [reply container/container_contract.go:113]: Fixed.

File: container/container_contract.go (hunk 3 above)
Comment (fyrchik) [reply container/container_contract.go:434]: I mirrored `List` logic here, but we can iterate over `ownerKeyPrefix` too, fixed.

File: container/container_contract.go (hunk 2 above)
Comment (fyrchik) [reply container/container_contract.go:430]: Fixed.

File: container/container_contract.go (hunk 1 above)
Comment (roman-khimov) [reply container/container_contract.go:98]: Not very user-friendly, upgrading contracts multiple times in no fun especially given that some node versions work fine with some contract versions and don't work with others.

File: container/container_contract.go [hunk 6]
```diff
@@ -700,29 +729,30 @@ func Version() int {
...
+	containerListKey = append(containerListKey, id...)
 	storage.Delete(ctx, containerListKey)
 
-	storage.Delete(ctx, id)
+	storage.Delete(ctx, append([]byte{containerKeyPrefix}, id...))
 }
 
 func getAllContainers(ctx storage.Context) [][]byte {
-	var list [][]byte
+	list := [][]byte{}
```
Comment (carpawell) [inline container/container_contract.go:749]: what's the byte code diff with that lines?

File: tests/container_test.go [hunk 7]
```diff
@@ -103,15 +103,56 @@ func TestContainerCount(t *testing.T) {
...
+
+		actual := make([][]byte, 0, len(expected))
+		for i := range arr {
+			id, ok := arr[i].Value().([]byte)
+			require.True(t, ok)
+			actual = append(actual, id)
+		}
+		require.ElementsMatch(t, expected, actual)
+	})
+	t.Run("check with `iterate`", func(t *testing.T) {
```
Comment (carpawell) [inline tests/container_test.go:139]:
```suggestion
	t.Run("check with `containersOf`", func(t *testing.T) {
```

**Review by carpawell — APPROVED** [review]

File: container/container_contract.go (hunk 6 above)
Comment (roman-khimov) [reply container/container_contract.go:749]: Should be none, which means that the change is not needed.

Comment (alexvanin) [reply container/container_contract.go:749]: I think empty `var list [][]byte` returns `[{"type":"Any"}]` on stack, while `list := [][]byte{}` returns `[{"type":"Array","value":[]}]`.

File: tests/container_test.go (hunk 7 above)
Comment (fyrchik) [reply tests/container_test.go]: Fixed

**Pull request merged**